src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```diff
@@ -0,0 +1,132 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+internal sealed class WebAssemblyDispatcher : Dispatcher
+{
+    public static readonly Dispatcher Instance = new WebAssemblyDispatcher();
+    private readonly SynchronizationContext? _context;
+
+    private WebAssemblyDispatcher()
+    {
+        // capture the JSSynchronizationContext from the main thread.
+        // if SynchronizationContext.Current is null on main thread, we are in single-threaded flavor of the dotnet wasm runtime
+        _context = SynchronizationContext.Current;
+    }
```

Comment:
Would it make sense to have `WebAssemblyDispatcher.Instance` be a factory method that creates either the multi-threaded dispatcher (this one) or a NullDispatcher for single-threaded wasm?

Reply:
done